docs/vcg-debugging.md

```diff
@@ -0,0 +1,124 @@
+
+```isabelle
+   ∀cte cap.
+      ccap_relation (cteCap cte) cap ⟶
+      cap_get_tag cap ≠ signed cap_vspace_cap ∧ ...
+```
+
+Which is obviously not provable, because we don't have any information about
+`cap` to work with. What it should have generated is something like:
+
+```isabelle
+   ∀cte cap.
+       ccap_relation (cteCap cte) cap ⟶
+       (∀t. cap_get_tag cap = signed cap_vspace_cap ⟶ ...)
+```
+
+This means that some conjunct in the `...` above got resolved to `False`, which
+got simplified to the unsolvable goal.
+
+The way to track down these issues is to go over the `vcg` steps to see which
+ones introduced the unexpected result. However, with large goals this is very
+difficult to work with. What else can help?
+
+First, try removing `(P ⟶ False) = (¬ P)` from the simpset, so we can look for
+`False` in the goals:
+
+```isabelle
+supply simp_thms(19)[simp del]
+```
+
+Second, if you do have identified a suspicious `vcg` invocation that could
+result in a bad instantiation, try the following instead of only `vcg`:
+
+```isabelle
+  apply (rule conseqPre, vcg, rule subset_refl)
+```
+
+For a "good" goal, this sequence will succeed and leave a goal to prove which
+has no schematic variables in the assumptions (ideally also not in the
+conclusion, but those should be harmless for this purpose). If this sequence
+fails, then there most likely is a problem with the parameters of the schematic
+precondition the goal is allowed to depend on, e.g. you have `?P x` in the
+precondition, but `vcg` is generating a goal that depends on add additional
+parameter `y`.
+
+This can happen when new parameters are introduced by `exE`, `clarsimp`, case
+distinctions, or other tactics. For instance you might have `x = Some y` in the
+assumptions and `?P (Some y)` as the schematic precondition, but the `vcg` needs
+`?P y`. Unification is generally not smart enough to figure out that `P (the
+(Some y))` would be a solution. The following strategies may help:
+
+* try using the tactic `wpfix`. It can sometimes resolve the problem by
+  replacing `?P` with a new `?P` that is allowed to depend on more parameters.
+
+* go back to the place where the schematic `?P` is introduced and introduce the
+  parameter before the schematic, so `?P` can depend on the parameter. This could
+  happen for instance by using a better case distinction rule that manages
+  schematics, or by performing the relevant `exE`, `clarsimp` etc before the
+  schematic is introduced (not always possible).
+
+* go back to the place where the parameter `y` is introduced and prevent it from
+  being introduced. E.g. instead of using `clarsimp` to generate `x = Some y`
+  and using `y`, be careful to keep the form `∃y. x = Some y` in your
+  assumptions (or prevent it from even occurring when possible, e.g. by blocking
+  a `x ≠ None` rewrite). Then use `the x` instead to access the content of `x`.
+  This is not necessarily nice, but should always be available in principle.
```

Comment:
should we mention `not_None` in `Lib.thy`?

Reply:
good idea, will add